### PR TITLE
[Backport] [GC] ZGC on AArch64

### PR DESCRIPTION
**Pull request by tanghaoth90**

### CODE REVIEW

**Review by linade — APPROVED** [review]

**Review by mmyxym — COMMENTED** [review]
Describe the difference in src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp
in summary.

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp [hunk 1]
```diff
@@ -1013,7 +1013,13 @@ void LIR_Assembler::mem2reg(LIR_Opr src, LIR_Opr dest, BasicType type, LIR_Patch
...
       __ decode_heap_oop(dest->as_register());
     }
-    __ verify_oop(dest->as_register());
+
+#if !INCLUDE_ZGC
+    if (!UseZGC) {
+      // Load barrier has not yet been applied, so ZGC can't verify the oop here
+      __ verify_oop(dest->as_register());
+    }
+#endif
```
Comment (mmyxym) [inline src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:1022]: #if
#else
 to guarantee the isolation?

Comment (tanghaoth90) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:1022]: thanks.

it's essentially add `if (!UseZGC) { .. }` before and after `__ verify_oop`. therefore, I modify it in the way below.

```
#if INCLUDE_ZGC
    if (!UseZGC) {
      // Load barrier has not yet been applied, so ZGC can't verify the oop here
#endif
      __ verify_oop(dest->as_register());
#if INCLUDE_ZGC
    }
#endif
```

**Review by mmyxym — APPROVED** [review]

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+                                   Address src,
+                                   Register tmp1,
+                                   Register tmp_thread) {
+  if (!ZBarrierSet::barrier_needed(decorators, type)) {
+    // Barrier not needed
+    BarrierSetAssembler::load_at(masm, decorators, type, dst, src, tmp1, tmp_thread);
+    return;
+  }
+
+  // rscratch1 can be passed as src or dst, so don't use it.
```
Comment (kuaiwei) [inline src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:66]: In openjdk master, rscratch1 can be used. Is there other patch fix it?
And can we use tmp1 here?

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+  __ mov(c_rarg1, rscratch2);
+
+  int step = 4 * wordSize;
+  __ mov(rscratch1, -step);
+  __ sub(sp, sp, step);
+
+  for (int i = 28; i >= 4; i -= 4) {
+    __ st1(as_FloatRegister(i), as_FloatRegister(i+1), as_FloatRegister(i+2),
+        as_FloatRegister(i+3), __ T1D, Address(__ post(sp, rscratch1)));
+  }
```
Comment (kuaiwei) [inline src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:100]: these code are used push/pop registers. I think there should be a common routine.

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+  // Verify value
+  if (type == T_OBJECT || type == T_ARRAY) {
+    // Note that src could be noreg, which means we
+    // are storing null and can skip verification.
+    if (val != noreg) {
+      Label done;
+
+      // tmp1 and tmp2 are often set to noreg.
+      RegSet savedRegs = RegSet::of(rscratch1);
+      __ push(savedRegs, sp);
```
Comment (kuaiwei) [inline src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:141]: do we need save rscratch1?

**Review by kuaiwei — APPROVED** [review]

**Review by kuaiwei — APPROVED** [review]

**Comment by tanghaoth90** [comment]
I will rearrange the relative patches, see https://github.com/alibaba/dragonwell11/issues/89

**Pull request closed without merging**